cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -379,17 +382,25 @@ protected void onFinish(HttpResponder responder, File uploadedFile) {
           ArtifactSummary artifactSummary = appRequest.getArtifact();
           NamespaceId artifactNamespace =
             ArtifactScope.SYSTEM.equals(artifactSummary.getScope()) ? NamespaceId.SYSTEM : appId.getParent();
-          Id.Artifact artifactId =
-            Id.Artifact.from(artifactNamespace.toId(), artifactSummary.getName(), artifactSummary.getVersion());
 
           KerberosPrincipalId ownerPrincipalId =
             appRequest.getOwnerPrincipal() == null ? null : new KerberosPrincipalId(appRequest.getOwnerPrincipal());
 
           // if we don't null check, it gets serialized to "null"
           String configString = appRequest.getConfig() == null ? null : GSON.toJson(appRequest.getConfig());
-          applicationLifecycleService.deployApp(appId.getParent(), appId.getApplication(), appId.getVersion(),
-                                                artifactId, configString, createProgramTerminator(), ownerPrincipalId,
-                                                appRequest.canUpdateSchedules());
+          if (artifactSummary.isVersionRange()) {
```

Comment:
as mentioned in other places, it's better not to have separate code paths for range and single version. I think you can just change applicationLifecycleService.deployApp() to take an ArtifactSummary instead of Id.Artifact, and let the lifecycle service determine the correct artifact.